### ISSUE TITLE
Head of xAH crashes when running jet systematics

### ISSUE THREAD

**Issue by johnda102** [UCATLAS/xAODAnaHelpers#236]
Head of xAH is broken.

The list of systemaitcs that it tries to run over is given below.. 
Looks like the Nominal is included more than once. So the code crashes. 

There were a lot of changes recently to getListOfSystematics I suspect this is where the problem is coming from.  Were these changes tested on jets ?  

JetCalibrator::initial... INFO     Running with systematic : 
JetCalibrator::initial... INFO     Running with systematic : 
JetCalibrator::initial... INFO     Running with systematic : JET_GroupedNP_1__1up
JetCalibrator::initial... INFO     Running with systematic : JET_GroupedNP_1__1down
JetCalibrator::initial... INFO     Running with systematic : JET_GroupedNP_2__1up
JetCalibrator::initial... INFO     Running with systematic : JET_GroupedNP_2__1down
JetCalibrator::initial... INFO     Running with systematic : JET_GroupedNP_3__1up
JetCalibrator::initial... INFO     Running with systematic : JET_GroupedNP_3__1down
JetCalibrator::initial... INFO     Running with systematic : 
JetCalibrator::initial... INFO     Running with systematic : JET_JER_SINGLE_NP__1up

**Comment by mmilesi** [comment]
Hi @johnda102 ,

I thought the full jet part regarding syst treatment had been updated by @gfacini, but I guess he checked only the BTag section so far. I have created a remote branch:

**fixJetSyst**

Which *might* fix the problem. However, is definitely unclear to me how to configure the JetCalibrator to run on systematics...can you please let me know how you did?

Btw, it looks like the systematic settings in JetCalibrator is quite different than in the other calibrators. Is there a way to make the whole thing more uniform? E.g., where is 'm_sysName` set?

Marco

**Comment by johnda102** [comment]
Hi Marco, 

   To get the systemaitcs I run with the config file given below with:

and calling 
setSyst( "All", 1.0 ); 
on the JetCalibrator

m_sysName gets set in the base class from the setSyst call above. 
Im not sure why the treatment is different here than in other classes. 
I agree we should harmonize in the future. 

config file:
InputContainer    AntiKt4EMTopoJets
JetAlgorithm      AntiKt4EMTopo
OutputContainer   AntiKt4EMTopoJets_Calib
OutputAlgo        AntiKt4EMTopoJets_Calib_Algo
Sort              true
SaveAllCleanDecisions true
# 13 TeV                                                                                                                                              
configNameAFII    JES_MC15Prerecommendation_AFII_June2015.config
configNameFullSim JES_MC15Prerecommendation_April2015.config
configNameData    JES_MC15Prerecommendation_April2015.config
# JetCalibrator adds _Insitu for the data                                                                                                             
CalibSequence     JetArea_Residual_Origin_EtaJES_GSC
#JESUncertConfig   $ROOTCOREBIN/data/JetUncertainties/JES_2015/Prerec/PrerecJES2015_3NP_Scenario1_50ns.config                                         
JESUncertConfig   $ROOTCOREBIN/data/JetUncertainties/JES_2015/Prerec/PrerecJES2015_3NP_Scenario1_25ns.config
JESUncertMCType   MC15
# JER                                                                                                                                                 
JERUncertConfig         JetResolution/Prerec2015_xCalib_2012JER_ReducedTo9NP_Plots_v2.root
JERFullSys        false
JERApplyNominal   false

**Comment by johnda102** [comment]
"# 13 TeV"

**Comment by mmilesi** [comment]
Hi @johnda102 ,

unfortunately I am not going to be able to test it today. Could you try to check out the branch I pushed to remote, and test it yourself? 

Cheers

Marco

**Comment by mmilesi** [comment]
Hi @johnda102 ,

now it works. Please look at the changes I made (very small indeed) and tell me if there's something unclear.

Marco

**Comment by johnda102** [comment]
Hi Marco, 

  ok great thanks.  For the moment I just reverted to an older version.  Ill try to get the head working again, maybe over the weekend. 

john

**Comment by kratsg** [comment]
Any updates on this? It would be great.

**Comment by btuan** [comment]
@kratsg 

Hi Giordon! I (and I believe @mswiatlo) have had luck working with the updates I made in my IncrementalFix branch. 

As far as I can tell, the systematics handling I implemented is in line with the handling of the other calibrators in xAH: nominal case is run unless "None" or a specific systematic is specified, and guards are strengthened to prevent the nominal case being added multiple times by different tools. 

The other major change is changing the config entry name from `OutputAlgo` to `OutputAlgoSystNames` for clarity and harmonization with the rest of the xAH calibrator classes, but this change can easily be reverted for backwards compatibility!

Best,
Brian

**Comment by kratsg** [comment]
I think you might as well start a PR.

**Comment by kratsg** [comment]
#241 has been merged. Can @johnda102 verify that this fix works?

**Comment by johnda102** [comment]
hi guys, 

 I will take a look at this today.

**Comment by johnda102** [comment]
So the head is still crashing for me in JetSelector when getting the systNames from the tstore. 

```c++
RETURN_CHECK("JetSelector::execute()", HelperFunctions::retrieve(systNames, m_inputAlgo, 0, m_store, m_verbose) ,"");
```

**Comment by btuan** [comment]
Hi @johnda102,

The config for the calibrator changed the OutputAlgo line item to
OutputAlgoSystNames, so perhaps nothing was recorded in the TStore in the
calibrator step.

Best,
Brian

PS this change seems to be a bit troublesome for backwards compatibility. I
can add a guard for it if this poses a problem

**Comment by johnda102** [comment]
Christ. OK let me change the config files and see if I can get it going.

**Comment by johnda102** [comment]
OK indeed that seems to have fixed it, however now when I try to run on all systematics eg: by giving '''-systAll 1 ''' in the submission I only seem to get the nominal jet branches... 

Did the syntax of configuring the JES/JER variations change ?

**Comment by johnda102** [comment]
looks like m_systName is being overwritten in 
```c++
EL::StatusCode  JetCalibrator :: configure ()
``` 

I think we need to use: 
```c++
    m_systName                = config->GetValue("SystName",            m_systName.c_str());
``` 
 instead of 

```c++
    m_systName                = config->GetValue("SystName",            "");
```

**Comment by jdandoy** [comment]
Hi @btuan @gfacini @kratsg @mmilesi ,

Sorry I missed the conversation about this previously, but I wanted to revisit this briefly.

It is always my preference to keep backwards compatibility, especially when it's easy to do.  Lots of people who don't pay attention to development use this tool.  All it requires is to add a second line:
    m_outputAlgo     = config->GetValue("OutputAlgoSystNames", m_outputAlgo.c_str());
    m_outputAlgo     = config->GetValue("OutputAlgo", m_outputAlgo.c_str()); //legacy option

Also this outputAlgo (currently called outputAlgoSystNames) doesn't have to have anything to do with systematics, so it's a bit misleading to have Syst in the name.   I'd like to revert this in a pull request (https://github.com/UCATLAS/xAODAnaHelpers/pull/255)

I also missed the boat on this, but I wouldn't think it would be a good idea to include the nominal case in "getSystematics()" .  First because the name is not truthful (getNominalAndSystematics).  Second is that  user algorithms may want to call more than 1 type of systematic, and now it requires them to keep track of how many getSystematics() calls have been made and to manually remove the nominal container for each subsequent call.  You can see these acrobatics that Marco had to make in this commit: https://github.com/UCATLAS/xAODAnaHelpers/commit/b09e8efbfc782764f5473d860eda4e2489d1e2f7  .   Here we have 2 types of systematics, Jet Energy Scale and Jet Energy Resolution.

Bottom line I don't think the convenience of adding nominal to getSystematics outweighs the added confusion.

**Comment by mmilesi** [comment]
Hi @btuan @gfacini @jdandoy 

Can we close this?
